ambry-router/src/main/java/com.github.ambry.router/AdaptiveOperationTracker.java

```diff
@@ -0,0 +1,211 @@
+   * @param crossColoEnabled {@code true} if requests can be sent to remote replicas, {@code false}
+   *                                otherwise.
+   * @param successTarget The number of successful responses required to succeed the operation.
+   * @param parallelism The maximum number of inflight requests at any point of time.
+   * @param time the {@link Time} instance to use.
+   * @param localColoTracker the {@link Histogram} that tracks intra datacenter latencies for this class of requests
+   * @param crossColoTracker the {@link Histogram} that tracks inter datacenter latencies for this class of requests
+   * @param quantile the quantile cutoff to use for when evaluating requests against the trackers.
+   */
+  AdaptiveOperationTracker(String datacenterName, PartitionId partitionId, boolean crossColoEnabled, int successTarget,
+      int parallelism, Time time, Histogram localColoTracker, Histogram crossColoTracker, double quantile) {
+    this.datacenterName = datacenterName;
+    this.time = time;
+    this.successTarget = successTarget;
+    this.parallelism = parallelism;
+    this.localColoTracker = localColoTracker;
+    this.crossColoTracker = crossColoTracker;
+    this.quantile = quantile;
+
+    // Order the replicas so that local healthy replicas are ordered and returned first,
+    // then the remote healthy ones, and finally the possibly down ones.
+    List<ReplicaId> replicas = partitionId.getReplicaIds();
+    LinkedList<ReplicaId> downReplicas = new LinkedList<>();
+    Collections.shuffle(replicas);
+
+    for (ReplicaId replicaId : replicas) {
+      String replicaDcName = replicaId.getDataNodeId().getDatacenterName();
+      if (!replicaId.isDown()) {
+        if (replicaDcName.equals(datacenterName)) {
+          replicaPool.addFirst(replicaId);
+        } else if (crossColoEnabled) {
+          replicaPool.addLast(replicaId);
+        }
+      } else {
+        if (replicaDcName.equals(datacenterName)) {
+          downReplicas.addFirst(replicaId);
+        } else if (crossColoEnabled) {
+          downReplicas.addLast(replicaId);
+        }
+      }
+    }
+    replicaPool.addAll(downReplicas);
+    totalReplicaCount = replicaPool.size();
+    if (totalReplicaCount < successTarget) {
+      throw new IllegalArgumentException(
+          "Total Replica count " + totalReplicaCount + " is less than success target " + successTarget);
+    }
+    this.otIterator = new OpTrackerIterator();
+  }
+
+  @Override
+  public boolean hasSucceeded() {
+    return succeededCount >= successTarget;
+  }
+
+  @Override
+  public boolean isDone() {
+    return hasSucceeded() || hasFailed();
+  }
+
+  @Override
+  public void onResponse(ReplicaId replicaId, boolean isSuccessFul) {
+    inflightCount--;
+    if (isSuccessFul) {
+      succeededCount++;
+    } else {
+      failedCount++;
+    }
+    getTracker(replicaId).update(time.milliseconds() - replicaRequestSendTimes.remove(replicaId));
```

Comment:
I did this to account for the NOT_FOUND failures which I want to factor in. Most failures probably do come in really early/late but since this has been piloted with the GET operations, I thought it would be imperative to account for their most time consuming failure.